In [38]:
import numpy as np
import random

Дан случайный массив, поменять знак у элементов, значения которых между 3 и 8

In [26]:
arr = np.random.randint(1, 20, size=10)
print ('Original array', arr)
mask = (arr >= 3) & (arr <= 8)
arr[mask] = -arr[mask]
print ('New array', arr)

Original array [ 7 11  1 10 10  8  6  3  4  8]
New array [-7 11  1 10 10 -8 -6 -3 -4 -8]


Заменить максимальный элемент случайного массива на 0

In [27]:
arr = np.random.randint(1, 20, size=10)
print ('Original array', arr)

# первый вариант
max_index = np.argmax(arr)
arr[max_index] = 0

# второй вариант
#max_value = arr.max()
#arr[arr == max_value] = 0

print ("New array", arr)

Original array [ 3 16 16 19  4 13  8  4 12  5]
New array [ 3 16 16  0  4 13  8  4 12  5]


Построить прямое произведение массивов (все комбинации с каждым элементом). На вход подается двумерный массив

In [28]:
def cartesian_product(arrays):
    arrays = [np.asarray(a) for a in arrays]
    shape = [len(a) for a in arrays]
    ix = np.indices(shape, dtype=int)
    ix = ix.reshape(len(arrays), -1).T
    result = np.empty(ix.shape, dtype=arrays[0].dtype)
    for n, a in enumerate(arrays):
        result[:, n] = a[ix[:, n]]
    return result

arr = np.array([
    [1, 2, 3],
    [4, 5],
    [6, 7]
], dtype=object)
arrays = list (arr)
result = cartesian_product(arr)
print(result)

[[1 4 6]
 [1 4 7]
 [1 5 6]
 [1 5 7]
 [2 4 6]
 [2 4 7]
 [2 5 6]
 [2 5 7]
 [3 4 6]
 [3 4 7]
 [3 5 6]
 [3 5 7]]


Даны 2 массива A (8x3) и B (2x2). Найти строки в A, которые содержат элементы из каждой строки в B, независимо от порядка элементов в B

In [33]:
np.random.seed(0) 
A = np.random.randint(0, 10, size=(8, 3))
B = np.random.randint(0, 10, size=(2, 2))

def find_rows_containing_each_B_row(A, B):
    result_indices = []
    for i, row_A in enumerate(A):
        set_A = set(row_A)
        contains_all = True
        for row_B in B:
            set_B = set(row_B)
            if not set_B.issubset(set_A):
                contains_all = False
                break
        if contains_all:
            result_indices.append(i)
    return result_indices

indices = find_rows_containing_each_B_row(A, B)

print("Массив A:")
print(A)
print("\nМассив B:")
print(B)
print("\nИндексы строк в A, которые содержат элементы из каждой строки B:")
print(indices)

Массив A:
[[5 0 3]
 [3 7 9]
 [3 5 2]
 [4 7 6]
 [8 8 1]
 [6 7 7]
 [8 1 5]
 [9 8 9]]

Массив B:
[[4 3]
 [0 3]]

Индексы строк в A, которые содержат элементы из каждой строки B:
[]


Дана 10x3 матрица, найти строки из неравных значений (например строка [2,2,3] остается, строка [3,3,3] удаляется)

In [20]:
Z = np.random.randint(0,5,(10,3))
E = np.logical_and.reduce(Z[:,1:] == Z[:,:-1], axis=1)
U = Z[~E]
print(Z)
print(U)

[[0 1 0]
 [1 1 1]
 [3 3 4]
 [1 3 1]
 [2 1 2]
 [3 0 1]
 [3 2 4]
 [2 1 2]
 [2 1 1]
 [0 2 1]]
[[0 1 0]
 [3 3 4]
 [1 3 1]
 [2 1 2]
 [3 0 1]
 [3 2 4]
 [2 1 2]
 [2 1 1]
 [0 2 1]]


Дан двумерный массив. Удалить те строки, которые повторяются

In [32]:
arr = np.random.randint(0, 10, size=(10, 3))

print("Исходный массив:")
print(arr)

# Удаление повторяющихся строк
unique_arr = np.unique(arr, axis=0)

print("Массив без повторяющихся строк:")
print(unique_arr)

Исходный массив:
[[8 0 8]
 [5 7 5]
 [7 1 9]
 [7 7 9]
 [3 0 6]
 [3 3 3]
 [4 4 3]
 [7 0 3]
 [6 0 1]
 [2 3 9]]
Массив без повторяющихся строк:
[[2 3 9]
 [3 0 6]
 [3 3 3]
 [4 4 3]
 [5 7 5]
 [6 0 1]
 [7 0 3]
 [7 1 9]
 [7 7 9]
 [8 0 8]]


Для каждой из следующих задач (1-5) нужно привести 2 реализации – одна без использования numpy (cчитайте, что там, где на входе или выходе должны быть numpy array, будут просто списки), а вторая полностью векторизованная с использованием numpy (без использования питоновских циклов/map/list comprehension).

**Замечание 1.** Можно считать, что все указанные объекты непустые (к примеру, в **задаче 1**на диагонали матрицы есть ненулевые элементы).

**Замечание 2.** Для большинства задач решение занимает не больше 1-2 строк.

**Задача 1:** Подсчитать произведение ненулевых элементов на диагонали прямоугольной матрицы.
Например, для X = np.array([[1, 0, 1], [2, 0, 2], [3, 0, 3], [4, 4, 4]]) ответ 3.

на списках (без numpy)

In [37]:
m, n = 4, 3
X = [[random.randint(0, 4) for _ in range(n)] for _ in range(m)]

def product_nonzero_diagonal(matrix):
    product = 1
    for i in range(min(len(matrix), len(matrix[0]))):
        if matrix[i][i] != 0:
            product *= matrix[i][i]
    return product

print("Матрица:")
for row in X:
    print(row)
print("Произведение ненулевых элементов на диагонали:", product_nonzero_diagonal(X))


Матрица:
[3, 1, 4]
[4, 4, 4]
[4, 2, 2]
[1, 3, 2]
Произведение ненулевых элементов на диагонали: 24


numpy

In [39]:
m, n = 4, 3
X = np.random.randint(0, 5, (m, n))

def product_nonzero_diagonal_np(X):
    diag = np.diagonal(X)
    nonzero_diag = diag[diag != 0]
    return np.prod(nonzero_diag)

print("Матрица:\n", X)
print("Произведение ненулевых элементов на диагонали:", product_nonzero_diagonal_np(X))

Матрица:
 [[0 2 3]
 [0 1 3]
 [3 3 0]
 [1 1 1]]
Произведение ненулевых элементов на диагонали: 1


**Задача 2:** Даны два вектора x и y. Проверить, задают ли они одно и то же мультимножество.
Например, для x = np.array([1, 2, 2, 4]), y = np.array([4, 2, 1, 2]) ответ True.

на списках (без numpy)

In [41]:
def check_multiset_equal(x, y):
    return sorted(x) == sorted(y)

x = [1, 2, 2, 4]
y = [4, 2, 1, 2]
print(check_multiset_equal(x, y))

True


numpy

In [42]:
def check_multiset_equal_np(x, y):
    return np.array_equal(np.sort(x), np.sort(y))

x = np.array([1, 2, 2, 4])
y = np.array([4, 2, 1, 2])
print(check_multiset_equal_np(x, y))

True


**Задача 3:** Найти максимальный элемент в векторе x среди элементов, перед которыми стоит ноль. Например, для x = np.array([6, 2, 0, 3, 0, 0, 5, 7, 0]) ответ 5.

без numpy (со списками)

In [44]:
def max_after_zero(x):
    candidates = [x[i] for i in range(1, len(x)) if x[i-1] == 0]
    return max(candidates) if candidates else None

x = [6, 2, 0, 3, 0, 0, 5, 7, 0]
print(max_after_zero(x))

5


numpy

In [45]:
def max_after_zero_np(x):
    zeros_before = (x[:-1] == 0)
    candidates = x[1:][zeros_before]
    return np.max(candidates) if candidates.size > 0 else None

x = np.array([6, 2, 0, 3, 0, 0, 5, 7, 0])
print(max_after_zero_np(x))

5


**Задача 4:** Реализовать кодирование длин серий (Run-length encoding). Для некоторого вектора x необходимо вернуть кортеж из двух векторов одинаковой длины. Первый содержит числа, а второй - сколько раз их нужно повторить.
Например, для x = np.array([2, 2, 2, 3, 3, 3, 5]) ответ (np.array([2, 3, 5]), np.array([3, 3, 1])).

без numpy (со списками)

In [46]:
def run_length_encoding(x):
    if not x:
        return [], []
    values = [x[0]]
    counts = [1]
    for i in range(1, len(x)):
        if x[i] == x[i-1]:
            counts[-1] += 1
        else:
            values.append(x[i])
            counts.append(1)
    return values, counts

x = [2, 2, 2, 3, 3, 3, 5]
values, counts = run_length_encoding(x)
print(values)  # [2, 3, 5]
print(counts)  # [3, 3, 1]

[2, 3, 5]
[3, 3, 1]


numpy

In [47]:
def run_length_encoding_np(x):
    x = np.asarray(x)
    if x.size == 0:
        return np.array([]), np.array([])
    # Находим индексы смены значения
    change = np.diff(x) != 0
    indices = np.append(np.where(change)[0], x.size - 1)
    counts = np.diff(np.append(-1, indices))
    values = x[indices]
    return values, counts

x = np.array([2, 2, 2, 3, 3, 3, 5])
values, counts = run_length_encoding_np(x)
print(values)  # [2 3 5]
print(counts)  # [3 3 1]

[2 3 5]
[3 3 1]


**Задача 5:** Даны две выборки объектов - X и Y. Вычислить матрицу евклидовых расстояний между объектами. Сравните с функцией scipy.spatial.distance.cdist по скорости работы.

без numpy (со списками)

In [51]:
def run_length_encoding(x):
    if not x:
        return [], []
    values, counts = [x[0]], [1]
    for i in range(1, len(x)):
        if x[i] == x[i-1]:
            counts[-1] += 1
        else:
            values.append(x[i])
            counts.append(1)
    return values, counts

x = [2, 2, 2, 3, 3, 3, 5]
v, c = run_length_encoding(x)
print(v)
print(c)

[2, 3, 5]
[3, 3, 1]


numpy

In [52]:
import numpy as np

def run_length_encoding_np(x):
    x = np.asarray(x)
    if x.size == 0:
        return np.array([]), np.array([])
    change = np.diff(x) != 0
    indices = np.append(np.where(change)[0], x.size - 1)
    counts = np.diff(np.append(-1, indices))
    values = x[indices]
    return values, counts

x = np.array([2, 2, 2, 3, 3, 3, 5])
v, c = run_length_encoding_np(x)
print(v)
print(c)

[2 3 5]
[3 3 1]


## Задача 6: CrunchieMunchies*
Вы работаете в отделе маркетинга пищевой компании MyCrunch, которая разрабатывает новый вид вкусных, полезных злаков под названием **CrunchieMunchies**

Вы хотите продемонстрировать потребителям, насколько полезны ваши хлопья по сравнению с другими ведущими брендами, поэтому вы собрали данные о питании нескольких разных конкурентов.

Ваша задача - использовать вычисления Numpy для анализа этих данных и доказать, что ваши **СrunchieMunchies** - самый здоровый выбор для потребителей.

1. Просмотрите файл cereal.csv. Этот файл содержит количества калорий для различных марок хлопьев. Загрузите данные из файла и сохраните их как calorie_stats.

In [60]:
calorie_stats = np.loadtxt("./data/cereal.csv", delimiter=",")
calorie_stats

array([ 70., 120.,  70.,  50., 110., 110., 110., 130.,  90.,  90., 120.,
       110., 120., 110., 110., 110., 100., 110., 110., 110., 100., 110.,
       100., 100., 110., 110., 100., 120., 120., 110., 100., 110., 100.,
       110., 120., 120., 110., 110., 110., 140., 110., 100., 110., 100.,
       150., 150., 160., 100., 120., 140.,  90., 130., 120., 100.,  50.,
        50., 100., 100., 120., 100.,  90., 110., 110.,  80.,  90.,  90.,
       110., 110.,  90., 110., 140., 100., 110., 110., 100., 100., 110.])

2. В одной порции CrunchieMunchies содержится 60 калорий. Насколько выше среднее количество калорий у ваших конкурентов?

Сохраните ответ в переменной average_calories и распечатайте переменную в терминале

In [61]:
average_calories = np.mean(calorie_stats)
print("Average calories of competitors:", average_calories)

Average calories of competitors: 106.88311688311688


3. Корректно ли среднее количество калорий отражает распределение набора данных? Давайте отсортируем данные и посмотрим.

Отсортируйте данные и сохраните результат в переменной calorie_stats_sorted. Распечатайте отсортированную информацию

In [62]:
calorie_stats_sorted = np.sort(calorie_stats)
print("Sorted calorie stats:", calorie_stats_sorted)

Sorted calorie stats: [ 50.  50.  50.  70.  70.  80.  90.  90.  90.  90.  90.  90.  90. 100.
 100. 100. 100. 100. 100. 100. 100. 100. 100. 100. 100. 100. 100. 100.
 100. 100. 110. 110. 110. 110. 110. 110. 110. 110. 110. 110. 110. 110.
 110. 110. 110. 110. 110. 110. 110. 110. 110. 110. 110. 110. 110. 110.
 110. 110. 110. 120. 120. 120. 120. 120. 120. 120. 120. 120. 120. 130.
 130. 140. 140. 140. 150. 150. 160.]


4. Похоже, что большинство значений выше среднего. Давайте посмотрим, является ли медиана наиболее корректным показателем набора данных.

Вычислите медиану набора данных и сохраните свой ответ в median_calories. Выведите медиану, чтобы вы могли видеть, как она сравнивается со средним значением.

In [63]:
median_calories = np.median(calorie_stats)
print("Median calories:", median_calories)

Median calories: 110.0


5. В то время как медиана показывает, что по крайней мере половина наших значений составляет более 100 калорий, было бы более впечатляюще показать, что значительная часть конкурентов имеет более высокое количество калорий, чем CrunchieMunchies.

Рассчитайте различные процентили и распечатайте их, пока не найдете наименьший процентиль, превышающий 60 калорий. Сохраните это значение в переменной nth_percentile.

In [64]:
percentiles = np.percentile(calorie_stats, np.arange(0, 101))
nth_percentile = None
for p in percentiles:
    if p > 60:
        nth_percentile = p
        break
print("Lowest percentile above 60 calories:", nth_percentile)

Lowest percentile above 60 calories: 70.0


6. Хотя процентиль показывает нам, что у большинства конкурентов количество калорий намного выше, это неудобная концепция для использования в маркетинговых материалах.

Вместо этого давайте подсчитаем процент хлопьев, в которых содержится более 60 калорий на порцию. Сохраните свой ответ в переменной more_calories и распечатайте его

In [65]:
more_calories = np.sum(calorie_stats > 60) / calorie_stats.size * 100
print("Percentage of cereals with calories > 60:", more_calories)

Percentage of cereals with calories > 60: 96.1038961038961


7. Это действительно высокий процент. Это будет очень полезно, когда мы будем продвигать CrunchieMunchies. Но один вопрос заключается в том, насколько велики различия в наборе данных? Можем ли мы сделать обобщение, что в большинстве злаков содержится около 100 калорий или разброс еще больше?

Рассчитайте величину отклонения, найдя стандартное отклонение, Сохраните свой ответ в calorie_std и распечатайте на терминале. Как мы можем включить эту ценность в наш анализ?

In [66]:
calorie_std = np.std(calorie_stats)
print("Standard deviation of calories:", calorie_std)

Standard deviation of calories: 19.35718533390827


8. Напишите короткий абзац, в котором кратко изложите свои выводы и то, как, по вашему мнению, эти данные могут быть использованы в интересах Mycrunch при маркетинге CrunchieMunchies.

1. **Среднее количество калорий** 
У конкурентов - 106.88 ккал на порцию
У CrunchieMunchies содержит всего 60 ккал (ниже среднего показателя)

2. **Медиана и распределение**
У конкурентов — 110.00 (половина брендов имеют калорийность выше 100)
Большинство конкурентов менее диетичны по сравнению с CrunchieMunchies

3. **Процентильный анализ**
У CrunchieMunchies минимальный процентиль превышает 60 калорий и равен 70.00 (у продукта низкая калорийность)
96.10% хлопьев содержат калорийность выше 60 (большинство конкурентов существенно калорийнее)

4. **Разброс данных**
Стандартное отклонение калорий 19.36 => разброс среди конкурентов c общим трендом в сторону более высокой калорийности 

**Вывод**

CrunchieMunchies — самый низкокалорийный и здоровый выбор среди рассмотренных вариантов
Это конкурентное преимущество стоит активно использовать в маркетинговых кампаниях, подчеркивая пользу и диетические преимущества CrunchieMunchies для здоровья потребителей